Gems/Atom/RPI/Code/Source/RPI.Edit/Material/MaterialTypeSourceData.cpp

```diff
@@ -112,89 +127,351 @@ namespace AZ
+                });
+
+            if (propertyGroupIter != m_propertyGroups.end())
+            {
+                AZ_Error("Material source data", false, "Property name '%.*s' collides with a PropertyGroup of the same name", AZ_STRING_ARG(name));
+                return nullptr;
+            }
+
+            if (!MaterialPropertyId::IsValidName(name))
+            {
+                AZ_Error("Material source data", false, "'%.*s' is not a valid identifier", AZ_STRING_ARG(name));
+                return nullptr;
+            }
+
+            m_properties.emplace_back(AZStd::make_unique<PropertyDefinition>(name));
+            return m_properties.back().get();
+        }
+
+        MaterialTypeSourceData::PropertyGroup* MaterialTypeSourceData::PropertyGroup::AddPropertyGroup(AZStd::string_view name)
+        {
+            auto iter = AZStd::find_if(m_properties.begin(), m_properties.end(), [name](const AZStd::unique_ptr<PropertyDefinition>& existingProperty)
+                {
+                    return existingProperty->GetName() == name;
+                });
+
+            if (iter != m_properties.end())
+            {
+                AZ_Error("Material source data", false, "PropertyGroup name '%.*s' collides with a Property of the same name", AZ_STRING_ARG(name));
+                return nullptr;
+            }
+
+            return AddPropertyGroup(name, m_propertyGroups);
+        }
+        
+        MaterialTypeSourceData::PropertyGroup* MaterialTypeSourceData::AddPropertyGroup(AZStd::string_view propertyGroupId)
+        {
+            AZStd::vector<AZStd::string_view> splitPropertyGroupId = SplitId(propertyGroupId);
+
+            if (splitPropertyGroupId.size() == 1)
+            {
+                return PropertyGroup::AddPropertyGroup(propertyGroupId, m_propertyLayout.m_propertyGroups);
+            }
+
+            PropertyGroup* parentPropertyGroup = FindPropertyGroup(splitPropertyGroupId[0]);
+            
+            if (!parentPropertyGroup)
+            {
+                AZ_Error("Material source data", false, "PropertyGroup '%.*s' does not exists", AZ_STRING_ARG(splitPropertyGroupId[0]));
+                return nullptr;
+            }
+
+            return parentPropertyGroup->AddPropertyGroup(splitPropertyGroupId[1]);
+        }
+        
+        MaterialTypeSourceData::PropertyDefinition* MaterialTypeSourceData::AddProperty(AZStd::string_view propertyId)
+        {
+            AZStd::vector<AZStd::string_view> splitPropertyId = SplitId(propertyId);
+
+            if (splitPropertyId.size() == 1)
+            {
+                AZ_Error("Material source data", false, "Property id '%.*s' is invalid. Properties must be added to a PropertyGroup (i.e. \"general.%.*s\").", AZ_STRING_ARG(propertyId), AZ_STRING_ARG(propertyId));
+                return nullptr;
+            }
+
+            PropertyGroup* parentPropertyGroup = FindPropertyGroup(splitPropertyId[0]);
+            
```

Comment:
Similar question, should this check if `splitPropertyId` is not empty before accessing its first element?